src/util.rs

```diff
@@ -0,0 +1,15 @@
+/// Utility functions.
+
+/// Returns true if a slice is a permutation.
+///
+/// That is, all the elements in `0..len` occur exactly once in the slice.
+pub fn is_permutation(v: &[usize]) -> bool {
+    let n = v.len();
+    let mut seen = (0..n).map(|_| false).collect::<Vec<_>>();
+    for &e in v {
+        if (0..n).contains(&e) {
+            seen[e] = true;
+        }
```

Comment:
Actually, since this can't be a valid permutation if it contains an out-of-bounds index, we could return immediately.

```suggestion
        if e >= n {
            return false;
        }
        seen[e] = true;
```